README.md
# Tencentcloud-Serverless-Nodejs

[![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](./LICENSE)
[![NPM Version](https://img.shields.io/npm/v/tencentcloud-serverless-nodejs.svg?style=flat)](https://www.npmjs.com/package/tencentcloud-serverless-nodejs)
[![NODE Version](https://img.shields.io/node/v/tencentcloud-serverless-nodejs.svg)](https://www.npmjs.com/package/tencentcloud-serverless-nodejs)


腾讯云云函数SDK，集成云函数业务流接口

## 配置需求
- 打开函数外网访问权限
- 如果没有手动传入secretId和secretKey等参数，函数需绑定有SCF Invoke 权限（或者包含SCF Invoke，比如SCF FullAccess）的角色

## Install
```shell
npm install tencentcloud-serverless-nodejs
```

## Example
```javascript
const { SDK, LogType }  = require('tencentcloud-serverless-nodejs')

exports.main_handler = async (event, context) => {
  context.callbackWaitsForEmptyEventLoop = false
  const sdk = new SDK({
    region:'ap-guangzhou'
  }) //如果在云函数中运行并且绑定了有SCF调用资格的运行角色，会默认取环境变量中的鉴权信息
  const res = await sdk.invoke({
    functionName: 'test_function',
    logType: LogType.Tail,
    data: {
      name: 'juli',
      role: 'big boss'
    }
  })
  console.log(res)
  // {
  //  "RequestId":"195434cd-ecf1-43d1-b707-90a293f5a282",
  //  "Result":{
  //     "BillDuration":3,
  //     "Duration":3,
  //     "ErrMsg":"",
  //     "FunctionRequestId":"195434cd-ecf1-43d1-b707-90a293f5a282",
  //     "InvokeResult":0,
  //     "Log":"START RequestId: 195434cd-ecf1-43d1-b707-90a293f5a282\nEvent RequestId: 195434cd-ecf1-43d1-b707-90a293f5a282\n2020-06-15T12:48:44.194Z\t195434cd-ecf1-43d1-b707-90a293f5a282\tHello World 111\n2020-06-15T12:48:44.194Z\t195434cd-ecf1-43d1-b707-90a293f5a282\t{ name: 'juli', role: 'big boss' }\n2020-06-15T12:48:44.196Z\t195434cd-ecf1-43d1-b707-90a293f5a282\tundefined\n2020-06-15T12:48:44.196Z\t195434cd-ecf1-43d1-b707-90a293f5a282\t{\n  callbackWaitsForEmptyEventLoop: [Getter/Setter],\n  getRemainingTimeInMillis: [Function: getRemainingTimeInMillis],\n  memory_limit_in_mb: 128,\n  time_limit_in_ms: 3000,\n  request_id: '195434cd-ecf1-43d1-b707-90a293f5a282',\n  environment: '{\"SCF_NAMESPACE\":\"default\"}',\n  environ: 'SCF_NAMESPACE=default;SCF_NAMESPACE=default',\n  function_version: '$LATEST',\n  function_name: 'test_function',\n  namespace: 'default',\n  tencentcloud_region: '',\n  tencentcloud_appid: 'xxxxxxxx',\n  tencentcloud_uin: 'xxxxxxxxxx'\n}\n\nEND RequestId: 195434cd-ecf1-43d1-b707-90a293f5a282\nReport RequestId: 195434cd-ecf1-43d1-b707-90a293f5a282 Duration:3ms Memory:128MB MemUsage:10.5977MB",
  //     "MemUsage":11112448,
  //     "RetMsg":"{\"name\":\"juli\",\"role\":\"big boss\"}"
  //  }
}
};
```

## API 

### Init

**new SDK(Params)**

使用SDK前，需要先初始化SDK，参数中undefined的值会被忽略。

**Params:**

| 参数名    | 是否必填 |  类型  |                                       描述 |
| :-------- | :------: | :----: | -----------------------------------------: |
| region    |    否    | string |  地域，默认广州（ap-guangzhou) |
| secretId  |    否    | string |  默认会取process.env.TENCENTCLOUD_SECRETID |
| secretKey |    否    | string | 默认会取process.env.TENCENTCLOUD_SECRETKEY |
| token |    否    | string | 默认会取process.env.TENCENTCLOUD_SESSIONTOKEN |


### Invoke

**invoke(Params)**

调用函数。支持同步和异步调用

**Params:**

| 参数名       | 是否必填 |  类型  |                    描述 |
| :----------- | :------: | :----: | ----------------------: |
| functionName |    是    | string |  函数名称 |
| qualifier    |    否    | string | 函数版本，默认为$LATEST |
| data         |    否    | string \| object |  函数运行入参 |
| namespace    |    否    | string | 命名空间，默认为default |
| invocationType |  否    | 'RequestResponse' \| 'Event' | 同步/异步调用
| routingKey   |    否    | string \| object |   函数灰度流量控制调用，以json格式传入，例如{"k":"v"}，注意kv都需要是字符串类型，最大支持的参数长度是1024字节
| logType      |    否    | 'None'\| 'Tail' | 日志类型，默认None，如传入Tail会返回调用过程中打印的日志

**Return:**

正常调用返回一个带有返回值结果和RequestId的object
``` json
 {
   "RequestId":"195434cd-ecf1-43d1-b707-90a293f5a282",
   "Result":{
      "BillDuration":3,
      "Duration":3,
      "ErrMsg":"",
      "FunctionRequestId":"195434cd-ecf1-43d1-b707-90a293f5a282",
      "InvokeResult":0,
      "Log":"START RequestId: 195434cd-ecf1-43d1-b707-90a293f5a282\nEvent RequestId: 195434cd-ecf1-43d1-b707-90a293f5a282\n2020-06-15T12:48:44.194Z\t195434cd-ecf1-43d1-b707-90a293f5a282\tHello World 111\n2020-06-15T12:48:44.194Z\t195434cd-ecf1-43d1-b707-90a293f5a282\t{ name: 'juli', role: 'big boss' }\n2020-06-15T12:48:44.196Z\t195434cd-ecf1-43d1-b707-90a293f5a282\tundefined\n2020-06-15T12:48:44.196Z\t195434cd-ecf1-43d1-b707-90a293f5a282\t{\n  callbackWaitsForEmptyEventLoop: [Getter/Setter],\n  getRemainingTimeInMillis: [Function: getRemainingTimeInMillis],\n  memory_limit_in_mb: 128,\n  time_limit_in_ms: 3000,\n  request_id: '195434cd-ecf1-43d1-b707-90a293f5a282',\n  environment: '{\"SCF_NAMESPACE\":\"default\"}',\n  environ: 'SCF_NAMESPACE=default;SCF_NAMESPACE=default',\n  function_version: '$LATEST',\n  function_name: 'test_function',\n  namespace: 'default',\n  tencentcloud_region: '',\n  tencentcloud_appid: 'xxxxxxxx',\n  tencentcloud_uin: 'xxxxxxxxxx'\n}\n\nEND RequestId: 195434cd-ecf1-43d1-b707-90a293f5a282\nReport RequestId: 195434cd-ecf1-43d1-b707-90a293f5a282 Duration:3ms Memory:128MB MemUsage:10.5977MB",
      "MemUsage":11112448,
      "RetMsg":"{\"name\":\"juli\",\"role\":\"big boss\"}"
   }
```

错误调用返回一个带有错误信息，错误码，以及RequestId的object，[【错误码详情】](https://cloud.tencent.com/document/product/583/30771)
```json
{
   "Error":{
      "Code":"ResourceNotFound.FunctionName",
      "Message":"Function not found: [n11112]"
   },
   "RequestId":"bb994f2f-526e-4142-bf9b-b8cf482c14ea"
}
```

## Licence

[MIT](./LICENSE)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.0.3"></a>
## [1.0.3](https://github.com/tencentyun/tencentcloud-serverless-nodejs/compare/v1.0.2...v1.0.3) (2020-10-13)



<a name="1.0.0"></a>
# [1.0.0](https://github.com/tencentyun/tencentcloud-serverless-nodejs/compare/v0.0.26...v1.0.0) (2020-06-16)


### Features

* 大版本升级，新增支持异步调用、灰度流量、输出log ([c1af7c5](https://github.com/tencentyun/tencentcloud-serverless-nodejs/commit/c1af7c5))


### BREAKING CHANGES

* 引入及调用API有修改，必须先初始化再调用，用new SDK()初始化sdk，调用不再接收secretId和secretKey参数，新增支持新功能的调用参数



<a name="0.0.26"></a>
## [0.0.26](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/compare/v0.0.25...v0.0.26) (2019-12-27)



<a name="0.0.25"></a>
## [0.0.25](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/compare/v0.0.24...v0.0.25) (2019-12-27)


### Bug Fixes

* **doc:** readme format ([f93d3e9](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/commit/f93d3e9))


### Features

* **init:** choose the internal endpoint when in tencentcloud serverless ([472267d](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/commit/472267d))



<a name="0.0.24"></a>
## [0.0.24](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/compare/v0.0.23...v0.0.24) (2019-11-14)


### Features

* **invoke:** support keep-alive ([23fabbd](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/commit/23fabbd))



<a name="0.0.23"></a>
## [0.0.23](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/compare/v0.0.22...v0.0.23) (2019-09-24)



<a name="0.0.22"></a>
## [0.0.22](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/compare/v0.0.21...v0.0.22) (2019-09-23)


### Features

* **init:** change throw error to warning when init without secretId/secretKey ([314b338](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/commit/314b338))


### BREAKING CHANGES

* **init:** change throw error to warning when init without secretId/secretKey



<a name="0.0.21"></a>
## [0.0.21](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/compare/v0.0.20...v0.0.21) (2019-08-22)



<a name="0.0.20"></a>
## [0.0.20](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/compare/v0.0.19...v0.0.20) (2019-08-22)


### Features

* **interface:invoke:** interface:invoke would get `secretId/secretKey/token` from `process.env` aut ([8ace71b](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/commit/8ace71b))



<a name="0.0.19"></a>
## [0.0.19](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/compare/v0.0.18...v0.0.19) (2019-08-21)



<a name="0.0.18"></a>
## [0.0.18](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/compare/v0.0.17...v0.0.18) (2019-08-19)



<a name="0.0.17"></a>
## [0.0.17](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/compare/v0.0.16...v0.0.17) (2019-08-19)



<a name="0.0.16"></a>
## [0.0.16](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/compare/v0.0.15...v0.0.16) (2019-08-16)


### Bug Fixes

* **interface:invoke:** fix interface:invoke argv ([f7314b2](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/commit/f7314b2))



<a name="0.0.15"></a>
## [0.0.15](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/compare/v0.0.14...v0.0.15) (2019-08-16)


### Features

* **interface:invoke:** add secretId/secretKey/token config to interface:invoke ([760e784](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/commit/760e784))



<a name="0.0.14"></a>
## [0.0.14](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/compare/v0.0.13...v0.0.14) (2019-08-16)



<a name="0.0.13"></a>
## [0.0.13](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/compare/v0.0.12...v0.0.13) (2019-07-28)


### Bug Fixes

* **ci:** resourced not found label fix ([87b0f8a](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/commit/87b0f8a))



<a name="0.0.12"></a>
## [0.0.12](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/compare/v0.0.11...v0.0.12) (2019-07-28)


### Bug Fixes

* **ts def fix:** ts def fix ([f72d7b3](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/commit/f72d7b3))



<a name="0.0.11"></a>
## [0.0.11](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/compare/v0.0.10...v0.0.11) (2019-05-08)



<a name="0.0.10"></a>
## [0.0.10](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/compare/v0.0.9...v0.0.10) (2019-05-08)



<a name="0.0.9"></a>
## [0.0.9](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/compare/v0.0.8...v0.0.9) (2019-05-08)


### Features

* **ci,api:** make invoke work without init ([fc3edb5](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/commit/fc3edb5))



<a name="0.0.8"></a>
## [0.0.8](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/compare/v0.0.7...v0.0.8) (2019-04-30)


### Bug Fixes

* **ci:** fix ci ([762b923](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/commit/762b923))



<a name="0.0.7"></a>
## [0.0.7](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/compare/v0.0.6...v0.0.7) (2019-04-30)



<a name="0.0.6"></a>
## [0.0.6](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/compare/v0.0.5...v0.0.6) (2019-04-30)



<a name="0.0.5"></a>
## [0.0.5](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/compare/v0.0.4...v0.0.5) (2019-04-30)


### Bug Fixes

* **invoke:** use sessionToken ([c31d3d7](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/commit/c31d3d7))



<a name="0.0.4"></a>
## [0.0.4](https://github.com/Lighting-Jack/tencentcloud-serverless-nodejs/compare/v0.0.3...v0.0.4) (2019-04-30)


### Bug Fixes


src/index.ts
import { Capi, CapiInstance } from '@tencent-sdk/capi'
import { InitConfig } from './helper/types'
import {
  MISSING_SECRET_ERROR,
  NOT_INIT_ERROR,
  REQUEST_ERROR
} from './helper/error'

export enum InvocationType {
  'RequestResponse' = 'RequestResponse', //同步调用（默认值）
  'Event' = 'Event' //异步调用
}

export enum LogType {
  'Tail' = 'Tail', //输出日志
  'None' = 'None' //不输出日志（默认值）
}

interface InvokeParam {
  /**函数名 */
  functionName: string
  /**版本/别名 */
  qualifier?: string
  /**运行函数的参数 */
  data?: string | object
  /**函数所在命名空间 */
  namespace?: string
  /**同步/异步调用 */
  invocationType?: InvocationType
  /**函数灰度流量控制调用，以json格式传入，例如{"k":"v"}，注意kv都需要是字符串类型，最大支持的参数长度是1024字节*/
  routingKey?: string | object
  /**是否输出日志 */
  logType?: LogType
}

export class SDK {
  public client: CapiInstance

  constructor(config?: InitConfig) {
    const defaultConfig = {
      secretId: process.env.TENCENTCLOUD_SECRETID,
      secretKey: process.env.TENCENTCLOUD_SECRETKEY,
      token: process.env.TENCENTCLOUD_SESSIONTOKEN,
      region: 'ap-guangzhou'
    }

    const mergedConfig = { ...defaultConfig, ...config }
    if (!mergedConfig.secretId || !mergedConfig.secretKey)
      throw new MISSING_SECRET_ERROR()
    this.client = new Capi({
      Region: mergedConfig.region,
      SecretId: mergedConfig.secretId,
      SecretKey: mergedConfig.secretKey,
      Token: mergedConfig.token,
      ServiceType: 'scf'
    })
  }


  _reset() {
    this.client = null
  }

  async invoke(invokeParam: InvokeParam) {
    if (!this.client) {
      throw new NOT_INIT_ERROR()
    }
    const res = await this.client.request(
      {
        Action: 'Invoke',
        Version: '2018-04-16',
        FunctionName: invokeParam.functionName,
        InvocationType: invokeParam.invocationType,
        Qualifier: invokeParam.qualifier,
        ClientContext: typeof invokeParam.data === 'string' ? invokeParam.data : JSON.stringify(invokeParam.data),
        Namespace: invokeParam.namespace,
        RoutingKey: typeof invokeParam.routingKey === 'string'? invokeParam.routingKey : JSON.stringify(invokeParam.routingKey),
        LogType: invokeParam.logType
      },
      {
        host: 'scf.tencentcloudapi.com'
      },
      true
    )
    if (res && res.Response) {
      return res.Response
    } else {
      throw new REQUEST_ERROR('Got no response:' + JSON.stringify(res))
    }
  }
}

export default SDK
